### DISCUSSION TITLE
Web failed to fetch worker logs in a multi datacenter setup

### DISCUSSION THREAD

**Discussion by miimsam** [apache/airflow#28219] [Q&A] [answered]
Hi There! We deployed our Airflow(Version 2.2.5) in two separate DCs and with Kubernetes Executor. Each DC has one scheduler pod and one web pod. Remote logging is enabled on S3, and we also set a custom logging class to write worker logs to stdout. 
The only problem with worker logs that we have is sometimes we get a "404 Not found" error on the log tab of a running task, something like the below one, even though I can see the worker pod is running:

```
*** Unable to fetch logs from worker pod exampleemailoperatorprintdate-b82ceb37bf1f4b32986cf058e3ab2baf ***
--
(404)
Reason: Not Found
HTTP response headers: HTTPHeaderDict({'Audit-Id': '2023ecd2-dafb-4520-ba37-2b0cfdd9a469', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'Date': 'Thu, 10 Nov 2022 13:54:10 GMT', 'Content-Length': '292'})
HTTP response body: b'{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"pods \\"exampleemailoperatorprintdate-b82ceb37bf1f4b32986cf058e3ab2baf\\" not found","reason":"NotFound","details":{"name":"exampleemailoperatorprintdate-b82ceb37bf1f4b32986cf058e3ab2baf","kind":"pods"},"code":404}\n'
 
```

My best guess is that the web instance can only fetch logs of workers running in the same DC (Kubernetes context) as it runs at the time and couldn't access the other DC pods. 

I wonder if we have missed anything in the configuration that caused this. Does anyone have any idea how we can overcome this issue?

**Comment by potiuk** [comment, marked as answer]
If yopu want to make the logs available accross the datacenters, you need to make the worker hostname/IP address available to be reached from the other datacenter..

In this case you need to provide a hostname_callable python function that will return the right IP Address (public?) or Hostname of the worker it runs on https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#hostname-callable

All the firewalls etc. need to be open for it to work, so your webserver from one Datacenter should be able to connect over the https to the worker in the other.

Alternatively - you might want to use cloudwatch instead of S3. The problem wit S3 that this is an object storage rather than log serving solution, so logs are uploaded to S3 only after the tasks are complete. S3 is not a regular file, you cannot append stuff to an object, you can upload the whole log in a single go, so S3 integration has the limitation that when the tasks are running, webserver will not be able to pull the log from S3 (because it's not there yet) so it will try to reach the worker. If everything is fine and your task will complete and upload the log to S3, my best guess is that you will start to see the logs regardless of the datacenter - because it will be downloaded from S3.

When you go "Cloudwatch" way on the other hand - cloudwatch is a fully-fledged log solution that supports both streaming the logs TO to cloudwatch from worker as well as streaming the logs as they appear from cloudwatch by the webserver. While it will have small delay (upload -> download), it should work regardless where your webservers/workers are located.
